File: src/services/satMode.ts
/**
 * Satellite Mode Handler (ap2016)
 *
 * Returns PLMN allow/barred lists and service constraints
 * for satellite connectivity mode when enabled.
 */

import type { ApplicationConfig } from '../protocol/responseTypes.js';
import { EntitlementStatus } from '../protocol/statusCodes.js';

export interface SatModeConfigData {
  plmnAllow?: string[];
  plmnBarred?: string[];
  serviceConstraints?: string;
}

export function buildSatModeConfig(
  status: number,
  provStatus: number,
  tcStatus: number,
  configData?: unknown,
): ApplicationConfig {
  const data = (configData ?? {}) as SatModeConfigData;

  const result: ApplicationConfig = {
    appId: 'ap2016',
    entitlementStatus: status,
    provStatus,
    tcStatus,
  };

  if (status === EntitlementStatus.ENABLED) {
    const extra: Record<string, string> = {};
    if (data.plmnAllow && data.plmnAllow.length > 0) extra.PLMNAllow = data.plmnAllow.join(',');
    if (data.plmnBarred && data.plmnBarred.length > 0) extra.PLMNBarred = data.plmnBarred.join(',');
    if (data.serviceConstraints) extra.ServiceConstraints = data.serviceConstraints;
    if (Object.keys(extra).length > 0) {
      result.extraParams = extra;
    }
  }

  return result;
}

File: internal/services/sat_mode.go
package services

import (
	"strings"

	"github.com/hkassaei/poc-entitlements-as-a-service-go/internal/protocol"
)

// SatModeConfigData holds configData for ap2016.
type SatModeConfigData struct {
	PLMNAllow          []string `json:"plmnAllow,omitempty"`
	PLMNBarred         []string `json:"plmnBarred,omitempty"`
	ServiceConstraints string   `json:"serviceConstraints,omitempty"`
}

// BuildSatModeConfig builds the ApplicationConfig for Satellite Mode (ap2016).
func BuildSatModeConfig(status, provStatus, tcStatus int, configData *SatModeConfigData) *protocol.ApplicationConfig {
	if configData == nil {
		configData = &SatModeConfigData{}
	}

	result := &protocol.ApplicationConfig{
		AppID:             "ap2016",
		EntitlementStatus: status,
		ProvStatus:        protocol.IntPtr(provStatus),
		TcStatus:          protocol.IntPtr(tcStatus),
	}

	if status == protocol.EntitlementStatusEnabled {
		extra := map[string]string{}
		if len(configData.PLMNAllow) > 0 {
			extra["PLMNAllow"] = strings.Join(configData.PLMNAllow, ",")
		}
		if len(configData.PLMNBarred) > 0 {
			extra["PLMNBarred"] = strings.Join(configData.PLMNBarred, ",")
		}
		if configData.ServiceConstraints != "" {
			extra["ServiceConstraints"] = configData.ServiceConstraints
		}
		if len(extra) > 0 {
			result.ExtraParams = extra
		}
	}

	return result
}
